File: tools/merge-content-block.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";

const [, , livePath, packagePath, targetIndexPath, targetTemplatePath] = process.argv;

if (!livePath || !packagePath || !targetIndexPath) {
  console.error(
    "Usage: merge-content-block.mjs <live-index> <package-index> <target-index> [target-template]",
  );
  process.exit(1);
}

const OPEN = '<script id="baked_data">';
const CLOSE = "</script>";

function readData(filePath) {
  const html = fs.readFileSync(filePath, "utf8");
  const start = html.indexOf(OPEN);
  const end = start === -1 ? -1 : html.indexOf(CLOSE, start);
  if (start === -1 || end === -1) {
    throw new Error(`Could not find the published content block in ${filePath}`);
  }
  const script = html.slice(start + OPEN.length, end).trim();
  const prefix = "window.__HALFSPACE_DATA__=";
  if (!script.startsWith(prefix) || !script.endsWith(";")) {
    throw new Error(`The published content block is malformed in ${filePath}`);
  }
  return JSON.parse(script.slice(prefix.length, -1));
}

function replaceData(filePath, data) {
  const html = fs.readFileSync(filePath, "utf8");
  const start = html.indexOf(OPEN);
  const end = start === -1 ? -1 : html.indexOf(CLOSE, start);
  if (start === -1 || end === -1) {
    throw new Error(`Could not update the published content block in ${filePath}`);
  }
  const block = `${OPEN}window.__HALFSPACE_DATA__=${JSON.stringify(data)};${CLOSE}`;
  fs.writeFileSync(
    filePath,
    html.slice(0, start) + block + html.slice(end + CLOSE.length),
  );
}

const liveData = readData(path.resolve(livePath));
const packageData = readData(path.resolve(packagePath));
const merged = structuredClone(liveData);

// These are interface/design records delivered by a structural update. All
// editorial content, rankings, positions, XIs, posts, and player records come
// from the newest published site instead.
const STRUCTURAL_KEYS = ["masthead_composer_v1"];
STRUCTURAL_KEYS.forEach((key) => {
  if (Object.prototype.hasOwnProperty.call(packageData, key)) {
    merged[key] = structuredClone(packageData[key]);
  }
});

merged.__content_revision_v1 =
  liveData.__content_revision_v1 || `installed-${new Date().toISOString()}`;
merged.__content_edit_clock_v1 = {};

replaceData(path.resolve(targetIndexPath), merged);
if (targetTemplatePath && fs.existsSync(targetTemplatePath)) {
  replaceData(path.resolve(targetTemplatePath), merged);
}

console.log(
  `Preserved ${Object.keys(liveData).length} published content sections; applied ${STRUCTURAL_KEYS.filter((key) => Object.prototype.hasOwnProperty.call(packageData, key)).length} structural setting.`,
);

File: tools/merge-content-block.py
#!/usr/bin/python3

import copy
import datetime
import json
import pathlib
import sys

if len(sys.argv) < 4:
    raise SystemExit(
        "Usage: merge-content-block.py <live-index> <package-index> <target-index> [target-template]"
    )

OPEN = '<script id="baked_data">'
CLOSE = "</script>"
PREFIX = "window.__HALFSPACE_DATA__="


def read_data(file_path):
    html = pathlib.Path(file_path).read_text(encoding="utf-8")
    start = html.find(OPEN)
    end = html.find(CLOSE, start)
    if start < 0 or end < 0:
        raise RuntimeError(f"Could not find the published content block in {file_path}")
    script = html[start + len(OPEN) : end].strip()
    if not script.startswith(PREFIX) or not script.endswith(";"):
        raise RuntimeError(f"The published content block is malformed in {file_path}")
    return json.loads(script[len(PREFIX) : -1])


def replace_data(file_path, data):
    target = pathlib.Path(file_path)
    html = target.read_text(encoding="utf-8")
    start = html.find(OPEN)
    end = html.find(CLOSE, start)
    if start < 0 or end < 0:
        raise RuntimeError(f"Could not update the published content block in {file_path}")
    payload = json.dumps(data, ensure_ascii=False, separators=(",", ":"))
    block = f"{OPEN}{PREFIX}{payload};{CLOSE}"
    target.write_text(html[:start] + block + html[end + len(CLOSE) :], encoding="utf-8")


live_data = read_data(sys.argv[1])
package_data = read_data(sys.argv[2])
merged = copy.deepcopy(live_data)
structural_keys = ["masthead_composer_v1"]

for key in structural_keys:
    if key in package_data:
        merged[key] = copy.deepcopy(package_data[key])

merged["__content_revision_v1"] = live_data.get(
    "__content_revision_v1"
) or "installed-" + datetime.datetime.now(datetime.timezone.utc).isoformat()
merged["__content_edit_clock_v1"] = {}

replace_data(sys.argv[3], merged)
if len(sys.argv) > 4 and pathlib.Path(sys.argv[4]).exists():
    replace_data(sys.argv[4], merged)

applied = sum(1 for key in structural_keys if key in package_data)
print(
    f"Preserved {len(live_data)} published content sections; applied {applied} structural setting."
)
